pkg/reconciler/buildrun/buildrun_test.go

```diff
@@ -547,89 +603,196 @@ var _ = Describe("Reconcile BuildRun", func() {
+						Reason: "ClusterBuildStrategyNotFound",
+						Status: corev1.ConditionFalse,
+					},
+					corev1.ConditionFalse,
+					buildSample.Spec,
+					true,
+				)
+				statusWriter.UpdateCalls(statusCall)
+
 				_, err := reconciler.Reconcile(buildRunRequest)
-				Expect(err).To(HaveOccurred())
-				Expect(err.Error()).To(ContainSubstring(" Unsupported BuildStrategy Kind"))
+				// we mark the BuildRun as Failed and do not reconcile again
+				Expect(err).ToNot(HaveOccurred())
 			})
 
-			It("only generates the service account once if the taskRun cannot be created", func() {
-				// override the Build to use a cluster BuildStrategy
+			It("fails on a TaskRun creation due to issues when retrieving the clusterbuildstrategy", func() {
+
+				// override the Build to use a namespaced BuildStrategy
 				buildSample = ctl.DefaultBuild(buildName, strategyName, build.ClusterBuildStrategyKind)
 
-				// override buildrun to use a generated service account
-				buildRunName = "foobar-buildrun-sa-generate"
-				buildRunSample = ctl.BuildRunWithSAGenerate(buildRunName, buildName)
-				buildRunSample.Labels = make(map[string]string)
-				buildRunSample.Labels[build.LabelBuild] = buildName
-				buildRunSample.Labels[build.LabelBuildGeneration] = strconv.FormatInt(buildSample.Generation, 10)
+				// stub get calls so that on cluster strategy retrieval, we throw a random error
+				stubGetCalls := func(context context.Context, nn types.NamespacedName, object runtime.Object) error {
+					switch object := object.(type) {
+					case *build.Build:
+						buildSample.DeepCopyInto(object)
+						return nil
+					case *build.BuildRun:
+						buildRunSample.DeepCopyInto(object)
+						return nil
+					case *corev1.ServiceAccount:
+						ctl.DefaultServiceAccount(saName).DeepCopyInto(object)
+						return nil
+					case *build.ClusterBuildStrategy:
+						return fmt.Errorf("something wrong happen")
+					}
+					return k8serrors.NewNotFound(schema.GroupResource{}, nn.Name)
+				}
 
-				// Override stub calls to include only build and buildrun
-				client.GetCalls(ctl.StubBuildRunGetWithoutSA(buildSample, buildRunSample))
+				client.GetCalls(stubGetCalls)
 
-				// Call the reconciler
 				_, err := reconciler.Reconcile(buildRunRequest)
-
-				// Expect an error stating that the strategy does not exist
+				// we reconcile again
 				Expect(err).To(HaveOccurred())
-				Expect(err.Error()).To(ContainSubstring(fmt.Sprintf(" \"%s\" not found", strategyName)))
+				Expect(err.Error()).To(ContainSubstring("something wrong happen"))
+				Expect(client.GetCallCount()).To(Equal(5))
+				Expect(client.StatusCallCount()).To(Equal(1))
+			})
 
-				// Expect one create call (service account)
-				Expect(client.CreateCallCount()).To(Equal(1))
-				_, obj, _ := client.CreateArgsForCall(0)
-				serviceAccount, castSuccessful := obj.(*corev1.ServiceAccount)
-				Expect(castSuccessful).To(BeTrue())
+			It("fails on a TaskRun creation due to unknown buildStrategy kind", func() {
```

Comment:
When the kind is missing, the fallback to the namespaced strategy is back. I think the unit test "only" fails to create the TaskRun because the namespaced build strategy does not exist. That's complicated to understand, I think. Would maybe be better if the test case would mock the namespaced build strategy to exist and then explicitly test this fallback.

Reply:
done